### ISSUE TITLE
Build Exception or Null Pointer Exception with various ic3.jar

### ISSUE THREAD

**Issue by CalixBarrus** [JordanSamhi/Tools#5]
Hi Jordan, I hope this is a good spot to reach you. 

I'm trying to get a working version of ic3. I've tried using some of the artifacts you've uploaded, specifically [this jar](https://github.com/JordanSamhi/Tools/blob/master/ic3.jar) from this Tools repo and [this one](https://github.com/JordanSamhi/RAICC/blob/master/artefacts/tools/ic3.jar) from your RAICC project. I've also tried building your fork of ic3 found on [this github branch](https://github.com/JordanSamhi/ic3). 

With the two jars from your Tools repo and RAICC repo I run into 

```
Exception in thread "main" java.lang.NullPointerException
        at soot.JastAddJ.Program.initPaths(Program.java:350)
        at soot.SootResolver.<init>(SootResolver.java:87)
        at soot.Singletons.soot_SootResolver(Singletons.java:802)
        at soot.SootResolver.v(SootResolver.java:91)
        at soot.Scene.tryLoadClass(Scene.java:670)
        at soot.Scene.loadBasicClasses(Scene.java:1268)
        at soot.Scene.loadNecessaryClasses(Scene.java:1347)
        at edu.psu.cse.siis.ic3.SetupApplication.initializeSoot(SetupApplication.java:343)
        at edu.psu.cse.siis.ic3.SetupApplication.calculateSourcesSinksEntrypoints(SetupApplication.java:124)
        at edu.psu.cse.siis.ic3.Ic3Analysis.initializeAnalysis(Ic3Analysis.java:146)
        at edu.psu.cse.siis.ic3.Ic3Analysis.initializeAnalysis(Ic3Analysis.java:64)
        at edu.psu.cse.siis.coal.Analysis.performAnalysis(Analysis.java:73)
        at edu.psu.cse.siis.ic3.Main.main(Main.java:13)
```
I'm not sure what could be causing this problem. Some [issues on the Soot repo with similar exceptions](https://github.com/soot-oss/soot/issues?q=soot.JastAddJ.Program.initPaths) discuss how java source code is not supported well by soot. This doesn't seem relevant, but maybe it has to do with some .class retargeting done for ic3 behind the scenes?

The build of your fork of ic3 runs into a different error:

```
Exception in thread "main" java.lang.NoClassDefFoundError: soot/javaToJimple/LocalGenerator
        at soot.jimple.infoflow.entryPointCreators.AndroidEntryPointCreator.createDummyMainInternal(AndroidEntryPointCreator.java:294)
        at soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator.createDummyMainInternal(BaseEntryPointCreator.java:123)
        at soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator.createDummyMain(BaseEntryPointCreator.java:108)
        at edu.psu.cse.siis.ic3.SetupApplication.createMainMethod(SetupApplication.java:313)
        at edu.psu.cse.siis.ic3.SetupApplication.calculateSourcesSinksEntrypoints(SetupApplication.java:125)
        at edu.psu.cse.siis.ic3.Ic3Analysis.initializeAnalysis(Ic3Analysis.java:146)
        at edu.psu.cse.siis.ic3.Ic3Analysis.initializeAnalysis(Ic3Analysis.java:64)
        at edu.psu.cse.siis.coal.Analysis.performAnalysis(Analysis.java:73)
        at edu.psu.cse.siis.ic3.Main.main(Main.java:13)
Caused by: java.lang.ClassNotFoundException: soot.javaToJimple.LocalGenerator
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 9 more
```

The error on this fork seems to be caused by a soot version mismatch. It looks like my build is using soot version 4.3.0. 

I'm running these experiments on Ubuntu 18.04, Java 11, Android SDKs 23 and 14.

**Comment by JordanSamhi** [comment]
Hi Calix,

Thanks for reporting these problems 👍 

Indeed the first problem seems to be related to Soot, not IC3 per see (note that IC3 runs with an old version of Soot. I had a mission last year to make it compatible with recent versions of Soot (and future ones), but it is a long and time-consuming task). Could you share the app you used to generate this exception? I will try it and see if I can do something locally. By the way, with my version of IC3 there is no retargeting anymore, it works directly with dex bytecode.

The second problem is definitely a problem with the Soot version. First, try recompiling with Java 8 :wink:  and let me know.

**Comment by CalixBarrus** [comment]
Using Java 8 was the solution for the two prebuilt JARs. Both ic3 jars in your Tools and RAICC repos work. This resolves my current issue. 

For documentation's sake, building your current branch of ic3 in Java 8 still produces a jar with the same error as above (NoClassDefFoundError). I was trying it with this apk from ICCBench, 
[icc_explicit_nosrc_nosink.apk.zip](https://github.com/JordanSamhi/Tools/files/15501296/icc_explicit_nosrc_nosink.apk.zip)

Running `mvn dependency:tree` in your ic3 branch gives in part

[INFO] edu.psu.cse.siis:ic3:jar:0.2.1
...
[INFO] +- infoflow:infoflow:jar:20150607:compile
[INFO] +- infoflow-android:infoflow-android:jar:20150607:compile
[INFO] +- com.google.protobuf:protobuf-java:jar:2.5.0:compile
**[INFO] \- com.jordansamhi:coal:jar:0.2.1:compile
[INFO]    +- com.jordansamhi:coal-strings:jar:0.2:compile
[INFO]    |  \- org.soot-oss:soot:jar:4.3.0:compile**
...

Thanks for your help!

**Comment by CalixBarrus** [comment]
I took another look at the runtime build error that I was getting when running [your branch of ic3](https://github.com/JordanSamhi/ic3/tree/master).

Exception in thread "main" java.lang.NoClassDefFoundError: soot/javaToJimple/LocalGenerator

The issue was that the pom.xml has a dependency on coal 0.2.1 which has been updated to a more recent version of soot. 
The 2015 infoflow code tries to find the renamed LocalGenerator class in the more recent soot version being accessed through the coal dependency. This error goes away when changing the dependency back to the 0.1.7 coal version included in ic3/libs.
